File: local_plugins/site24x7_flutter_plugin/site24x7_flutter_plugin-1.0.0/ios/Classes/ApmMobileapmFlutterPlugin.swift
import Flutter
import UIKit
import APM

public class ApmMobileapmFlutterPlugin: NSObject, FlutterPlugin {
    
    var s247TransactionsDictionary: [String : S24Transaction] = [:]
    var s247ComponentsDictionary: [String: S24Component] = [:]
    
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "apm_mobileapm_flutter_plugin", binaryMessenger: registrar.messenger())
    let instance = ApmMobileapmFlutterPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    
    let arguments = call.arguments as? [String: Any?]
      
    switch call.method {
        
      case "startMonitoring": do {
          let appKey = arguments?["appKey"] as? String
          let uploadInterval = arguments?["uploadInterval"] as? Int
          if(appKey != nil && uploadInterval != nil){
              S24APM.start(withAppKey: appKey, interval: UInt(uploadInterval!))
          }else{
              NSLog("site24x7 either appKey or uploadInterval might be nil")
          }
      }
      case "startMonitoringWithEndPoint": do {
          let appKey = arguments?["appKey"] as? String
          let uploadInterval = arguments?["uploadInterval"] as? Int
          let endPoint = arguments?["endPoint"] as? String
          if(appKey != nil && uploadInterval != nil && endPoint != nil){
              S24APM.start(withAppKey: appKey, interval: UInt(uploadInterval!), server: endPoint)
          }else{
              NSLog("site24x7 either appKey, uploadInterval or endPoint might be nil")
          }
      }
      
      case "startTransaction": do {
          let transactionName = arguments?["name"] as? String
          guard transactionName != nil, let s247Transaction = S24APM.startTransaction(withName: transactionName) else {
              NSLog("site24x7 name should not be nil")
              return
          }
          s247TransactionsDictionary[transactionName!] = s247Transaction
      }
      
      case "stopTransaction": do {
          let transactionName = arguments?["name"] as? String
          guard transactionName != nil, s247TransactionsDictionary.contains(where: {$0.key == transactionName!}) else {
              NSLog("site24x7 either name has been nil or transaction has not been started")
              return
          }
          let transaction = s247TransactionsDictionary.removeValue(forKey: transactionName!)
          S24APM.stop(transaction)
      }
          
      case "startComponent": do {
          let transactionName = arguments?["transactionName"] as? String
          let componentName = arguments?["componentName"] as? String
          if(transactionName != nil && componentName != nil && s247TransactionsDictionary.contains(where: {$0.key == transactionName!})){
              let transaction = s247TransactionsDictionary[transactionName!]
              let component = transaction?.startComponent(withType: transactionName!+componentName!)
              s247ComponentsDictionary["transactionName+componentName"] = component
          }else{
              NSLog("site24x7 either transaction, component names are nil or transaction has not been started")
          }
      }
          
      case "stopComponent": do {
          let transactionName = arguments?["transactionName"] as? String
          let componentName = arguments?["componentName"] as? String
          guard transactionName != nil, componentName != nil, s247TransactionsDictionary.contains(where: {$0.key == transactionName}), s247ComponentsDictionary.contains(where: {$0.key == (transactionName! + componentName!)}), let transaction = s247TransactionsDictionary[transactionName!], let component = s247ComponentsDictionary[transactionName!+componentName!] else {
              NSLog("site24x7 either transaction, component names are nil or transaction, component has not been started")
              return
          }
          transaction.stop(component)
      }
          
      case "crashNative": do {
          fatalError("This is a site24x7 crash")
      }
          
      case "flush": do {
          S24APM.flush()
      }
          
      case "setUserId": do {
          let userId = arguments?["userId"] as? String
          S24APM.setUserId(userId)
      }
          
      case "addBreadcrumb": do {
          let name = arguments?["name"] as? String
          let action = arguments?["action"] as? String
          if(name != nil && action != nil){
              S24APM.addBreadcrumb(with: name, andAction: action)
          }else{
              NSLog("site24x7 either name or action is nil")
          }
      }
          
      case "addScreen": do {
          let screenName = arguments?["name"] as? String
          let loadTime = arguments?["loadTime"] as? Int
          let startTime = arguments?["startTime"] as? Int
          if(loadTime != nil && startTime != nil && screenName != nil){
            S24APM.addScreen(screenName, withTime: UInt((loadTime!)), startTime: String(startTime!))
          }else{
              NSLog("site24x7 either loadTime, startTime or screenName is nil")
          }
      }
          
      case "addHttpCall": do {
          let url = arguments?["url"] as? String
          let requestMethod = arguments?["requestMethod"] as? String
          let startTime = arguments?["startTime"] as? Int
          let loadTime = arguments?["loadTime"] as? UInt
          let statusCode = arguments?["statusCode"] as? UInt
          let screen = arguments?["currentRoute"] as? String
          if(url != nil && requestMethod != nil && startTime != nil && loadTime != nil && statusCode != nil){
            S24APM.addHttpCall(url, withTime: loadTime!, startTime: String(startTime!), respCode: statusCode!, httpMethod: requestMethod)
          }else{
              NSLog("site24x7 either of url, requestMethod, startTime, loadTime or status code is nil")
          }
      }
          
      case "captureException": do {
          let msg = arguments?["msg"] as? String
          let file = arguments?["file"] as? String
          let function = arguments?["function"] as? String
          let startTime = arguments?["startTime"] as? Int
          let stack  = arguments?["stack"] as? String
          let screenName = arguments?["currentRoute"] as? String
          let type = arguments?["type"] as? String
          let isHandled = arguments?["isHandled"] as? Bool
          guard msg != nil, file != nil, function != nil, startTime != nil, stack != nil, screenName != nil, type != nil else{
              NSLog("site24x7 either of msg, file, function, start time, stack, screenName or type is nil")
              return
          }
          S24APM.addError(msg, errorType: type, file: file, methodName: function, stackTrace: stack, screen: screenName, startTime: String(startTime!), handledException: isHandled!)
      }
        
    case "trackNativeExceptions": do {
        S24APM.enableErrorReporting()
    }
      
    default:
        result(FlutterMethodNotImplemented)
    }
  }
}

File: local_plugins/site24x7_flutter_plugin/site24x7_flutter_plugin-1.0.0/android/src/main/java/com/zoho/apm_mobileapm_flutter_plugin/ApmMobileapmFlutterPlugin.java
package com.zoho.apm_mobileapm_flutter_plugin;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

import com.site24x7.android.apm.Apm;
import com.site24x7.android.apm.Component;
import com.site24x7.android.apm.Transaction;

import java.net.MalformedURLException;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

/** ApmMobileapmFlutterPlugin */
public class ApmMobileapmFlutterPlugin implements FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private MethodChannel channel;

  private Context context;
  private final Map<String, Transaction> transactionsMap = new HashMap<String,Transaction>();
  private Map<String, Component> componentsMap = new HashMap<String,Component>();

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "apm_mobileapm_flutter_plugin");
    channel.setMethodCallHandler(this);
    context = flutterPluginBinding.getApplicationContext();
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {

    switch (call.method) {

      case "startMonitoring": {
        String appKey = Objects.requireNonNull(call.argument("appKey")); // no I18N
        int uploadInterval = Integer.parseInt(Objects.requireNonNull(call.argument("uploadInterval")).toString());
        Apm.startMonitoring(this.context, appKey, uploadInterval);
        break;
      }

      case "startMonitoringWithEndPoint": {
        String appKey = Objects.requireNonNull(call.argument("appKey")); // no I18N
        int uploadInterval = Integer.parseInt(Objects.requireNonNull(call.argument("uploadInterval")).toString());
        String endPoint = Objects.requireNonNull(call.argument("endPoint"), "endpoint should not be null"); //no I18N
        try {
          Apm.startMonitoring(this.context, appKey, uploadInterval, endPoint);
        } catch (MalformedURLException e) {
          Log.d("site24x7", "exception while start monitoring", e); //no I18N
          throw new RuntimeException(e);
        }
        break;
      }

      case "trackNativeExceptions": {
        Apm.enableErrorReporting();
        break;
      }

      case "startTransaction": {
        try{
          String transactionName = Objects.requireNonNull(call.argument("name")); // no I18N
          Transaction transaction = Apm.startTransaction(transactionName);
          transactionsMap.put(transactionName, transaction);
        }catch (Exception e){
          Log.d("site24x7","Exception while starting transaction ", e); //no I18N
        }
        break;
      }

      case "stopTransaction": {
        String transactionName = Objects.requireNonNull(call.argument("name"), "transaction name should not be null"); // no I18N
        try{
          if (transactionsMap.containsKey(transactionName)) {
            Apm.stopTransaction(transactionsMap.remove(transactionName));
          }
        }catch (Exception e){
          Log.d("site24x7","Exception while stopping transaction ", e); //no I18N
        }
        break;
      }

      case "crashNative": {
        throw new RuntimeException("This is a site24x7 crash"); // no I18N
      }

      case "startComponent":{
        String txnName = Objects.requireNonNull(call.argument("transactionName"), "transaction name should not be null"); // no I18N
        String compName = Objects.requireNonNull(call.argument("componentName"), "component name should not be null"); // no I18N
        try{
          if(transactionsMap.containsKey(txnName)){
            componentsMap.put(txnName+compName,transactionsMap.get(txnName).startComponent(compName));
          }
        }catch (Exception e){
          Log.d("site24x7","exception while starting component ", e); //no I18N
        }
        break;
      }

      case "stopComponent": {
        String txnName = Objects.requireNonNull(call.argument("transactionName"), "transaction name should not be null"); // no I18N
        String compName = Objects.requireNonNull(call.argument("componentName"), "component name should not be null"); // no I18N
        try{
          if(transactionsMap.containsKey(txnName) && componentsMap.containsKey(txnName+compName)){
            transactionsMap.get(txnName).stopComponent(componentsMap.get(txnName + compName));
          }
        }catch (Exception e){
          Log.d("site24x7","exception while stopping component ", e); //no I18N
        }
        break;
      }

      case "flush": {
        try{
          Apm.flush();
        }catch (Exception e){
          Log.d("site24x7", "exception while flushing data", e); //no I18N
        }
        break;
      }

      case "setUserId": {
        String userId = Objects.requireNonNull(call.argument("userId"), "userId must not be null"); //no I18N
        try{
          Apm.setUserId(userId);
        }catch (Exception e){
          Log.d("site24x7", "exception while setting user id", e); //no I18N
        }
        break;
      }

      case "addBreadcrumb": {
        String name  = Objects.requireNonNull(call.argument("name")); //no I18N
        String action = Objects.requireNonNull(call.argument("action")); //no I18N
        try{
          Apm.addBreadcrumb(name, action);
        }catch (Exception e){
          Log.d("site24x7", "exception while adding breadcrumb", e); //no I18N
        }
        break;
      }

      case "addScreen": {
        String screenName = Objects.requireNonNull(call.argument("name"), "screen name should not be null"); //no I18N
        int loadTime = Objects.requireNonNull(call.argument("loadTime"), "load time should not be null"); //no I18N
        long startTime = Objects.requireNonNull(call.argument("startTime"), "start time should not be null"); //no I18N
        try{
          Apm.addScreen(screenName, (long) loadTime, startTime);
        }catch (Exception e){
          Log.d("site24x7", "exception while adding screen", e); //no I18N
        }
        break;
      }

      case "captureException": {
        String msg = Objects.requireNonNull(call.argument("msg"), "msg should not be null"); //no I18N
        String file = Objects.requireNonNull(call.argument("file"), "file should not be null"); //no I18N
        String function = Objects.requireNonNull(call.argument("function"), "function should not be null"); //no I18N
        long startTime = Objects.requireNonNull(call.argument("startTime"), "start time should not be null"); //no I18N
        String stack = Objects.requireNonNull(call.argument("stack"), "stack should not be null"); //no I18N
        String screenName = Objects.requireNonNull(call.argument("currentRoute"), "currentRoute should not be null"); //no I18N
        String type = Objects.requireNonNull(call.argument("type")); //no I18N
        try{
          Apm.addError(msg, type, file, function, stack, screenName, startTime);
        }catch (Exception e){
          Log.d("site24x7", "exception while adding error to dependent native agent"); //no I18N
        }
        break;
      }

      case "addHttpCall": {
        String url = Objects.requireNonNull(call.argument("url"), "url should not be null"); //no I18N
        String requestMethod = Objects.requireNonNull(call.argument("requestMethod"), "request method should not be null"); //no I18N
        long startTime = Objects.requireNonNull(call.argument("startTime"), "start time should not be null"); //no I18N
        int loadTime = Objects.requireNonNull(call.argument("loadTime"), "load time should not be null"); //no I18N
        int statusCode = Objects.requireNonNull(call.argument("statusCode"), "status code should not be null"); //no I18N
        String screen = Objects.requireNonNull(call.argument("currentRoute"), "current route should not be null"); //no I18N
        try{
          Apm.addHttpCalls(url, requestMethod, startTime, loadTime, statusCode, screen);
        }catch (Exception e){
          Log.d("site24x7", "exception while adding http call", e); //no I18N
        }
        break;
      }

      default:{
        result.notImplemented();
        break;
      }
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }
}
